### PR TITLE
add support for using xorriso to create the cdrom iso

### PR DESCRIPTION
**Pull request by rgl**
Related to https://github.com/hashicorp/packer/pull/9796

### CODE REVIEW

**Review by nywilken — CHANGES_REQUESTED** [review]
HI @rgl thanks for adding in support for xorriso. It is a good addition for sure.

Overall this is looking good. I tested on Debian with xorriso and it worked as expected, but I ran into issues after uninstalling both xorriso and removing my symlink of mkisofs to genisoimge. Previously Packer would error if no command was found, but now it crashes. I left a comment on where the fix needs to happen. 

I also left a couple of questions and minor suggestions. Let me know if you have any questions on my feedback.

File: common/step_create_cdrom.go
Comment (rgl) [reply common/step_create_cdrom.go:119]: Indeed, `xorriso` does not need `genisoimage` installed, its just a mode of operation of `xorriso`.

About the macOS, if there is a `xorriso` command maybe the user really wants to use it because it has to be explicitly installed. But I'm OK with moving `xorriso` to the end of the list; just let me known.

File: common/step_create_cdrom.go
Comment (rgl) [reply common/step_create_cdrom.go]: Please check it now.

File: common/step_create_cdrom.go
Comment (rgl) [reply common/step_create_cdrom.go]: Please check it now.

File: common/step_create_cdrom.go
Comment (nywilken) [inline common/step_create_cdrom.go:181]: Nice! Fix confirmed

```
Warning: Warning when preparing build: "ubuntu-2004-vmware"

A shutdown_command was not specified. Without a shutdown command, Packer
will forcibly halt the virtual machine, which may result in data loss.


ubuntu-2004-vmware: output will be in this color.

==> ubuntu-2004-vmware: Retrieving ISO
==> ubuntu-2004-vmware: Trying https://releases.ubuntu.com/20.04/ubuntu-20.04.1-live-server-amd64.iso
==> ubuntu-2004-vmware: Trying https://releases.ubuntu.com/20.04/ubuntu-20.04.1-live-server-amd64.iso?checksum=sha256%3A443511f6bf12402c12503733059269a2e10dec602916c0a75263e5d990f6bb93
==> ubuntu-2004-vmware: https://releases.ubuntu.com/20.04/ubuntu-20.04.1-live-server-amd64.iso?checksum=sha256%3A443511f6bf12402c12503733059269a2e10dec602916c0a75263e5d990f6bb93 => /home/wilken/Development/packer-templates/vmware-iso/vmware-iso-cdfiles/packer_cache/e30b34e4d3c536763868a42bb9d5b8f14af297b8.iso
==> ubuntu-2004-vmware: Configuring output and export directories...
==> ubuntu-2004-vmware: Creating CD disk...
==> ubuntu-2004-vmware: Deleting output directory...
Build 'ubuntu-2004-vmware' errored after 2 seconds 968 milliseconds: could not find a supported CD ISO creation command (the supported commands are: xorriso, mkisofs, hdiutil, oscdimg)

==> Wait completed after 2 seconds 968 milliseconds

==> Some builds didn't complete successfully and had errors:
--> ubuntu-2004-vmware: could not find a supported CD ISO creation command (the supported commands are: xorriso, mkisofs, hdiutil, oscdimg)

==> Builds finished but no artifacts were created.

```

**Review by nywilken — APPROVED** [review]
The updates looks great and work as expected. Thanks for the quick turnaround. I left one more suggestion to consolidate the loops in the `retrieveCDISOCreationCommand` function. But this looks like it is otherwise good to go. I would like to get @SwampDragons review before merging.

File: common/step_create_cdrom.go
Comment (SwampDragons) [reply common/step_create_cdrom.go:125]: I think it's fine as a default.

**Review by SwampDragons — APPROVED** [review]

**Comment by SwampDragons** [comment]
This looks good to me 👍

File: common/step_create_cdrom.go
Comment (rgl) [reply common/step_create_cdrom.go:183]: I have mix feelings about this change. As-is the array is never created when the user has a supported command, and the code seems simpler.

Comment (nywilken) [reply common/step_create_cdrom.go:183]: I might've honed in too much on the double for loop on this one. So I'm happy that you shared your thoughts. Seeing this "the code seems simpler" is the true north star for me :+1:

**Comment by nywilken** [comment]
Merging now that we got the SwampDragon's approval. Thanks!

**Comment by rgl** [comment]
Oh this is falling on me on Windows when running inside an MSYS2 bash:

```
==> windows-2019-amd64-hyperv: Creating CD disk...
    windows-2019-amd64-hyperv: GNU xorriso 1.5.0 : RockRidge filesystem manipulator, libburnia project.
    windows-2019-amd64-hyperv: Drive current: -outdev 'stdio:C:\Users\rgl\AppData\Local\Temp\packer773584472.iso'
    windows-2019-amd64-hyperv: Media current: stdio file, overwriteable
    windows-2019-amd64-hyperv: Media status : is blank
    windows-2019-amd64-hyperv: Media summary: 0 sessions, 0 data blocks, 0 data,  168g free
    windows-2019-amd64-hyperv: xorriso : FAILURE : Cannot determine attributes of source file '/c/rgl/windows-vagrant/C:\Users\rgl\AppData\Local\Temp\packer_to_cdrom871483351' : No such file or directory
    windows-2019-amd64-hyperv: xorriso : aborting : -abort_on 'FAILURE' encountered 'FAILURE'
==> windows-2019-amd64-hyperv: Bad exit status: 5
==> windows-2019-amd64-hyperv: Step "StepCreateCD" failed, aborting...
```

The logs have:

```
2020/09/16 19:54:55 packer.exe plugin: CD path: C:\Users\rgl\AppData\Local\Temp\packer773584472.iso
2020/09/16 19:54:55 packer.exe plugin: Wrote 8814 bytes to autounattend.xml
2020/09/16 19:54:55 packer.exe plugin: Wrote 3160 bytes to winrm.ps1
2020/09/16 19:54:55 packer.exe plugin: Executing: C:\tools\msys64\usr\bin\xorriso.exe [-as genisoimage -rock -joliet -volid autounattend -output C:\Users\rgl\AppData\Local\Temp\packer773584472.iso C:\Users\rgl\AppData\Local\Temp\packer_to_cdrom871483351]
```

There's a mix-up with the paths when they go from a win32 process (packer) to a msys2/cygwin process (xorriso).

To fix this I have to use some ugly heuristics to known whether I'm using a msys2/cygwin xorriso.exe:

1. packer is running on Windows.
2. the `xorriso.exe` absolute path ends with `\usr\bin\xorriso.exe`.
3. there's a `cygpath.exe` executable in $PATH
4. call `cygpath.exe` to convert from windows path to msys2/cygwin path.
4.1. in this case calling `C:\tools\msys64\usr\bin\cygpath.exe C:\Users\rgl\AppData\Local\Temp\packer773584472.iso` would return `/c/Users/rgl/AppData/Local/Temp/packer773584472.iso`
4.2. this works because cygpath.exe is a native win32 app.
5. sorry :-(

What do you think?

**Comment by rgl** [comment]
I've now implemented something similar to the above in #9954 and its working nicely :-)

**Pull request merged**
